pkg/plumbing/config/exec_config.go

```diff
@@ -0,0 +1,113 @@
+package config
+
+import (
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/warptools/warpforge/pkg/formulaexec"
+	"github.com/warptools/warpforge/pkg/plotexec"
+	"github.com/warptools/warpforge/pkg/workspace"
+	"github.com/warptools/warpforge/wfapi"
+)
+
+// ExecConfig is warpforge's internal configuration required to run Exec
+// Consumers should not accept this struct directly but should accept an interface for the parts that they need.
+type ExecConfig struct {
+	State
+	// WorkspaceSet is the workspace stack set to be used when executing.
+	// This controls behavior such as memoization and host warehouses.
+	workspace.WorkspaceSet
+	// This is the root filesystem as an fs.FS
+	Fsys fs.FS
+}
+
+func (c *ExecConfig) BinPath() string {
+	// determine the path of the running executable
+	// other binaries (runc, rio) will be located here as well
+	path, ok := c.State.Env[EnvWarpforgePath]
+	if ok {
+		return path
+	}
+	return filepath.Dir(c.State.ExecutablePath)
+}
+
+func (c *ExecConfig) KeepRunDir() bool {
+	return c.envIsSet(EnvWarpforgeKeepRundir)
+}
+
+func (c *ExecConfig) RunPathBase() string {
+	if value, ok := c.State.Env[EnvWarpforgeRunPath]; ok {
+		return value
+	}
+	return c.TempDir
+}
+
+func (c *ExecConfig) WarehousePathOverride() *string {
+	value, ok := c.State.Env[EnvWarpforgeWarehouse]
+	if !ok {
+		return nil
+	}
+	return &value
+}
+
+func (c *ExecConfig) WorkingDirectory() string {
+	return c.State.WorkingDirectory
+}
+
+// NewConfig returns a warpforge configuration
+//
+// Errors:
+//
+//   - warpforge-error-searching-filesystem -- unable to load workspace stack
+//   - warpforge-error-serialization -- failed to get copy of global state
+func NewConfig() (*ExecConfig, error) {
+	state, err := NewState()
+	if err != nil {
+		return &ExecConfig{}, err
+	}
+	fsys := os.DirFS("/")
+	wss, err := workspace.FindWorkspaceStack(fsys, "", state.WorkingDirectory[1:])
+	if err != nil {
+		return &ExecConfig{}, err
+	}
+	return &ExecConfig{
+		State:        state,
+		WorkspaceSet: wss,
+		Fsys:         fsys,
+	}, nil
+}
+
+func (c *ExecConfig) envIsSet(key string) bool {
+	_, result := c.State.Env[key]
+	return result
+}
+
+func (c *ExecConfig) envOverride(key string) (string, bool) {
+	a, b := c.State.Env[key]
+	return a, b
+}
+
+func PlotExecConfig(cfg *ExecConfig) *plotexec.ExecConfig {
+	return &plotexec.ExecConfig{
+		BinPath:          cfg.BinPath(),
+		KeepRunDir:       cfg.KeepRunDir(),
+		RunPathBase:      cfg.RunPathBase(),
+		WhPathOverride:   cfg.WarehousePathOverride(),
+		WorkingDirectory: cfg.WorkingDirectory(),
+		WorkspaceSet:     cfg.WorkspaceSet,
```

Comment:
So I'm must wonder out loud here: how many fields being flipped from struct to struct is it before it begins to be a smell?  It seems like here (and the block below) there's more fields being copied than fields that are novel in the new struct.

I'd seriously consider just joining more things together in one big struct.  A heuristic I'm using for how much I like this code is "if we find there's another piece of config to track across the system, how many lines am I going to have to add?" and "if I rename something, how many sites have the same name and deserve the same change, while being technically different sites?", and these struct copies seem likely to make both those numbers go up.  Those numbers going up mean ease of change is decreasing.

In the other direction, putting more things into one bulky struct is creating closer coupling, but... I don't think we really have a compelling reason to treat that as a negative in this situation.  We really do use approximately all of these values approximately everywhere in Warpforge, afaict.  Or to put it another way: I currently have no goalposts in mind where things would be sufficient not-close-coupled to be a worthy target -- so nudging in that direction, without a decisive goalpost, doesn't seem worth it to me.

Reply:
I tend to care less about how many lines might change and more about whether or not I can track inputs from clearly defined API points. Renaming some fields is one of the least annoying things in the set of annoying things related to writing software. This particular bit of code is still a bit questionable just because it's superfluous. Where we could handle it via the utility functions.